### ISSUE TITLE
Missing Inferred Type in Tooltip when using Visible Type Applications

### ISSUE THREAD

**Issue by julmb** [haskell/haskell-language-server#4017]
### Your environment

Which OS do you use?
ArchLinux

Which version of GHC do you use and how did you install it?
9.6.4 from ghcup

How is your project built (alternative: link to the project)?
cabal

Which LSP client (editor/plugin) do you use?
Code - OSS + vscode-haskell

Which version of HLS do you use and how did you install it?
2.6.0.0 from ghcup

Have you configured HLS in any way (especially: a `hie.yaml` file)?
```
"haskell.manageHLS": "PATH",
"haskell.plugin.importLens.codeActionsOn": false,
"haskell.plugin.class.codeActionsOn": false,
"haskell.plugin.semanticTokens.globalOn": true,
```

### Steps to reproduce

Hovering over `print` in this example:
```hs
main = (print :: Int -> IO ()) 0
```
![image](https://github.com/haskell/haskell-language-server/assets/11631652/c0d3fc8d-29b3-4f84-b44a-b9597a49f002)

Hovering over `print` in this example:
```hs
main = print @Int 0
```
![image](https://github.com/haskell/haskell-language-server/assets/11631652/2473dc66-2d8d-4bfd-9524-b49fa52c43b3)

### Expected behaviour

The inferred type for `print` is shown in the tooltip in both examples.

### Actual behaviour

The inferred type for `print` is only shown with type annotations via `::`, but not with visible type applications via `@`. The inferred type is shown correctly without any type annotations (together with a warning about defaulting the type variable to `Integer`).

### Debug information

I could not find anything relevant in the HLS output log either when typing the expression or when hovering to produce the tooltip.

**Comment by jhrcek** [comment]
I find this curious and managed to reproduce the issue.

I only got this far:
The types info is generated from the list of types retrieved from HieAST here:
https://github.com/haskell/haskell-language-server/blob/a29d8e8bd8264ab2bdcfab60de2521d39869e99a/ghcide/src/Development/IDE/Spans/AtPoint.hs#L293

When type application is not used, this usually contains 2 values (most general type and currently inferred type(?))
But when type application is present, this list of types is empty.
But I have no idea why. If only there was an easy way to pretty-print hie ast around the point of interest to see what's going on. I might look into this over the weekend just to learn more deeply how this stuff works..

**Comment by soulomoon** [comment]
They are pretty dependent on the hieAst.

`main = (print :: Int -> IO ()) 0` we have 

```
[(0, ()), (1, IO), (2, Int), (3, Show), (4, 'Many),
 (5, HFunTy 4 2 1), (6, Type), (7, HTyVarTy a), (8, Show),
 (9, HFunTy 4 7 1), (10, HQualTy 8 9),
 (11, HForAllTy ((a, 6), [spec]) 10)]
 
            Node@hello.hs:5:9-13: Source: From source
                                  {(annotations: {(HsVar, HsExpr), (XExpr, HsExpr)}), 
                                   (types: [5, 11]), 
                                   (identifier info: {(name print,  Details:  Just 11 {usage}), 
                                                      (name $dShow_aRb,  Details:  Just 3 {usage of evidence variable})})}
```

`main = print @Int 0` we have 
```
[(0, ()), (1, IO), (2, Int), (3, Show), (4, 'Many),
 (5, HFunTy 4 2 1), (6, Type), (7, HTyVarTy a), (8, Show),
 (9, HFunTy 4 7 1), (10, HQualTy 8 9),
 (11, HForAllTy ((a, 6), [spec]) 10)]
 
          Node@hello.hs:5:8-12: Source: From source
                                {(annotations: {(HsVar, HsExpr)}),  (types: [11]), 
                                 (identifier info: {(name print,  Details:  Just 11 {usage})})}
```

**Comment by jhrcek** [comment]
Nice investigation @soulomoon  I'd be interested how you got the printout of the relevant ranges.
And do you have to recompile/reinstall hls just to get the "print" statements in?

**Comment by soulomoon** [comment]
You can setup ghc session and build ast in a simple file  @jhrcek 
```
    let filePath = "hello.hs"
    content <- Prelude.readFile filePath
    runGhc (Just libdir) $ do
        dflags <- getSessionDynFlags
        let dflags' = gopt_set dflags Opt_Haddock
        setSessionDynFlags dflags'
        target <- guessTarget filePath Nothing Nothing
        setTargets [target]
        load LoadAllTargets
        modSum <- getModSummary $ mkModuleName "Hello"
        parsedModule <- parseModule modSum
        typecheckedModule <- typecheckModule parsedModule
        let rms = fromJust $ renamedSource typecheckedModule
        let tcGblEnv = fst $ tm_internals_ typecheckedModule
        file <- mkHieFile modSum tcGblEnv rms
        let ast = snd $ List.head $ Map.toList $ getAsts $ hie_asts file
        let types = Arr.elems $ hie_types file
        liftIO $ Prelude.writeFile "hieAst.txt" $ showSDoc fakeDynFlags (ppr (Prelude.zip [0::Int ..] types)) ++ "\n" ++ showSDoc fakeDynFlags (ppr ast)
```

**Comment by jhrcek** [comment]
Awesome, thanks! I'll play around with this snippet.

But I wonder what do the hls devs do when they want to investigate "what's in the hie ast". Isn't there a way to pretty print it for debugging purposes? You know, to debug things like "the info I was expected is not in this ast node"..

**Comment by soulomoon** [comment]
> Awesome, thanks! ~I'll play around with this snippet.~ I turned your snipped into runnable script for local poking around. It required me to declare Outputable instance for HieType to make it work. https://gist.github.com/jhrcek/9785e1e4ae909995d06a8810e1d6d98c
> 
> This seems pretty tedious. I wonder what do the hls devs do when they want to investigate "what's in the hie ast"? Isn't there a way to pretty print it for debugging purposes? You know, to debug things like "the info I was expecting is not in this ast node, where can I get it instead?"..

I am not sure the others, I just going through the source code of hls or ghc. it's a bit off topic here, maybe we can discuss this in matrix.

**Comment by michaelpj** [comment]
So it seems like this is a bug in HIE file generation? I would expect the `print` node to still have a type even with the type application.

**Comment by jhrcek** [comment]
I figured out a nicer way to show hie file (with full types reconstructed instead of with `TypeIndex`es) and it seems that in both cases the type info is associated with the `print` AST node. So not sure what leads to it not being rendered.
Will try to dig deeper over weekend.

![Screenshot from 2024-01-26 13-15-59](https://github.com/haskell/haskell-language-server/assets/2716069/cf4f15df-be8c-4a62-b968-a08096b19732)
![Screenshot from 2024-01-26 13-15-01](https://github.com/haskell/haskell-language-server/assets/2716069/1ed19b8a-c4c1-45a7-b513-4feeafbab16a)

**Comment by soulomoon** [comment]
> I figured out a nicer way to show hie file (with full types reconstructed instead of with `TypeIndex`es) and it seems that in both cases the type info is associated with the `print` AST node. So not sure what leads to it not being rendered. Will try to dig deeper over weekend.

Take a look at the evidence variables. They are the one with more concrete type.

**Comment by jhrcek** [comment]
Ok, I got to the bottom of this (what a fun learning experience this was :smile:)

TL;DR: not showing types when TypeApplications is used seems like deliberate design - the type is not shown at the bottom of markdown block because it is already shown at the top (so having it repeated at the bottom would seem redundant)


More details:
As I commented above (refer to the screenshots) the HieAST does contain type information in both cases:
1. when explicit type annotation is used there are 2 types in the `print`'s AST node:
- `forall a. Show a => a -> IO ()`
- `Int -> IO ()`
2. when TypeApplications is used there is only one type:
- `forall a. Show a => a -> IO ()`

Then there is this [dropEnd1 logic](https://github.com/haskell/haskell-language-server/blob/4788bfca11c366a8d184da2d13f12302429abd57/ghcide/src/Development/IDE/Spans/AtPoint.hs#L227) which basically drops the last type from the list in cases when there's only single identifier associated with the hovered-over symbol. That logic has been implemented back in 2020 in this PR: https://github.com/haskell/haskell-language-server/commit/5dfee4c8c90a5ac66d5a5a66b869e5596e52c1db#diff-91f773bae93bbf86035573f4dbffa3035c1689b1bb7e17bde5937bb1271c0131R104
I think the reason it's there is because that last type is actually rendered at the top of the markdown block at the top, so it seems it just "works as designed".

I hope the following screenshots from my debugging sessions (where I just `Debug.Trace`d the relevant expressions in the linked code) help to clarify the difference in behavior more then lengthy explanations:

![Screenshot from 2024-01-28 07-42-37](https://github.com/haskell/haskell-language-server/assets/2716069/6ec26452-5ac4-4f9d-896d-0997fae4f3d2)
![Screenshot from 2024-01-28 07-42-04](https://github.com/haskell/haskell-language-server/assets/2716069/5e08b92d-f1fb-4468-8b0b-259e4a812690)

**Comment by michaelpj** [comment]
Shame there's no comment explaining, maybe @wz1000 remembers and could write one?

It seems like the logic we _want_ is to not reproduce the "original" type signature. It's not clear that we're really achieving that, though; in that case we should be filtering out the original type signature from the list rather than what we're doing now. I'd be somewhat in favour of just displaying everything (ideally with a title for that section of the hover explaining what it means!), at least that's consistent.

**Comment by julmb** [comment]
This explains how the HIE AST is shown in the user interface, but I am still wondering why the AST contains only the general type `forall a. Show a => a -> IO ()` and not the instantiated one `Int -> IO ()`.

I was expecting HLS to behave similarly to GHCI in this case:
```
ghci> :type print
print :: Show a => a -> IO ()
ghci> :type print @Int
print @Int :: Int -> IO ()
```

In fact, I would argue that showing the instantiated signature is more important than showing the original one (which can always be looked up in the documentation or source).

Additionally, the fact that the user is using visible type applications indicates that the type inference situation is tricky and these are exactly the situations in which the user really benefits from insight into which types their expressions ultimately got assigned.

**Comment by wz1000** [comment]
This is just how the language works.

for `print (1 :: Int)`, the typechecker has to elaborate `print` to `print @Int`. Since `@Int` wasn't written explicitly by the user, it is attached to the `print`.

If the user explicitly writes `print @Int`, then the typechecker doesn't need to do any elaboration, and the type of the span that `print` is used in is `forall a. ...`. It would be incorrect to give that span the type `Int -> ...` because you can't apply a function `Int -> ...` to a type argument `@Int`.

I suppose what you want is the type of the expression `print @Int`, which is the feature request in #709, currently blocked on both GHC not providing us information in an easily accessible way and LSP not really supporting a reasonable way to expose this to the user.

**Comment by michaelpj** [comment]
I think there's maybe an issue in how people think about things indeed. I think it's quite natural to think that in `print (1 :: Int)` the typechecker uses the information from `(1 :: Int)` to infer that `print` "has" the type `Int -> ...`. That's not what actually happens: `print` always has the type `forall a . ...`, and the compiler just infers the type _argument_ to `print`. And indeed those come apart in the `print @Int` case. But I can see how it's confusing.

Perhaps one issue in the `print (1 :: Int)` case is that we report the type of (effectively) a non-existent expression, which is confusing. Is there a way we could _say_ that that's what we're reporting? in an ideal world I imagine something like having this in the hover:
```
print @Int :: Int -> ...
```

**Comment by julmb** [comment]
Ah, I think I understand now, that actually makes a lot of sense, thank you for explaining.

So what I really want is to view the type of the expression `print @Int`, which is currently not possible. And the only reason it works in the annotated case is that the typechecker elaborates `print` to `print @Int` and since the `@Int` is implicit, by hovering over `print`, I get the type of the expression `print @Int`.

The situation in the annotated case seems a little confusing, but nonetheless I find it incredibly useful to be able to view the inferred/elaborated types that way.

Of course, ideally, we would like to be able to view the types of arbitrary expressions as proposed in #709. I have been teaching Haskell to a friend of mine and I think that this functionality would be very useful for people learning the language, as many things make so much more sense when you see the underlying types of the expressions involved.

**Comment by michaelpj** [comment]
Thinking about this, I think there are some things we could do. In particular, while showing people the type of an arbitrary expression is hard, we can make some guesses about what expressions people might want to see. In particular, when hovering over an identifier here are a few ideas for things we could show:
1. The type of the identifier applied to any invisible arguments that it is supplied with. This would handle the `print @Int` case, and in general would mean we behave consistently whether invisible arguments are provided or not.
2. The type of the identifier applied to each prefix of arguments that it is provided with. This is going to be overkill in a lot of situations where there are long applications but might be interesting.
3. The type of the entire application that the identifier appears in. I think this would cover a bunch of the cases where people want the type of an expression. Perhaps more generally we could make a guess at the next "enclosing expression" and give the type of that?

**Comment by xsebek** [comment]
> It seems like the logic we want is to not reproduce the "original" type signature. It's not clear that we're really achieving that, though;

@michaelpj here is a case with more than one type argument where it does not work that way:

```Haskell
someFunc :: IO ()
someFunc = mapM_ @[] @IO print [1, 2, 3]
```

Without type annotations, you get the generic type and specialized type at the bottom:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/cae94f61-efbb-4000-8d76-401f152aa8cc" />

Once you add any type arguments, this information is missing, even if it would be useful to know the type arguments that were not provided:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/3c0d765d-2239-476a-bb8f-c1f188fe2760" />

I used examples with just Prelude types so this may seem silly. I would like type applications to work well with HLS, since they are easier to write than the full signature when types get tricky, e.g. in effect systems where specifying some type variables can help.

--- 

EDIT: I found that if you point at the last `@` it shows what I would like to see in the function hover:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/b95e0c95-9c94-4ca1-9290-98cf73a7fb02" />

**Comment by xsebek** [comment]
For future reference, the script above changed with GHC 9.14 ([ghc 5f213bff](https://gitlab.haskell.org/ghc/ghc/-/commit/5f213bffa87ed78100c5e64d80678bfb58c44d79?page=2)) - you need to wrap it in `Module`:
```diff
-        modSum <- getModSummary $ mkModuleName "Hello"
+        let mn = mkModuleName "Main"
+        let m  = mkModule (RealUnit (Definite (homeUnitId_ dflags))) mn
+        modSum <- getModSummary m
```
Or there is a tool `hiedb` which can just dump the AST:
```sh
cabal install hiedb
# add ghc-options: -fwrite-ide-info -hiedir .
hiedb dump Main.hie
```

**Comment by xsebek** [comment]
I created GHC issue to ask how HLS should deal with type applications: https://gitlab.haskell.org/ghc/ghc/-/issues/27339
